mesa/datacollection.py

```diff
@@ -151,28 +155,31 @@ def _new_table(self, table_name, table_columns):
         new_table = {column: [] for column in table_columns}
         self.tables[table_name] = new_table
 
-    def _record_agents(self, model):
+    def _record_agents(self, model, schedule):
         """Record agents data in a mapping of functions and agents."""
-        rep_funcs = self.agent_reporters.values()
-        if all(hasattr(rep, "attribute_name") for rep in rep_funcs):
-            prefix = ["model.schedule.steps", "unique_id"]
-            attributes = [func.attribute_name for func in rep_funcs]
-            get_reports = attrgetter(*prefix + attributes)
-        else:
-
-            def get_reports(agent):
-                _prefix = (agent.model.schedule.steps, agent.unique_id)
-                reports = tuple(rep(agent) for rep in rep_funcs)
-                return _prefix + reports
-
-        agent_records = map(get_reports, model.schedule.agents)
+        agent_records = map(
+            partial(self._get_reports, self, schedule.steps), schedule.agents
+        )
         return agent_records
 
+    @staticmethod
+    def _get_reports(collector, steps, agent):
```

Comment:
This method name is too generic. It could have meant getting the model report.

Reply:
It's not used elsewhere in the code. Why not put the function back in `_record_agents`?

Reply:
I wrongly assumed that by avoiding redefining the same function every call we'd get a performance improvement. This was not the case:

I used the introductory tutorial at ["Collecting Data"](https://mesa.readthedocs.io/en/latest/tutorials/intro_tutorial.html#collecting-data) and removed the `model_reporters`, then timed this:
```python
from timeit import default_timer as timer # not called while timing, just here to show what package I was using
start = timer()
model.datacollector.collect(model)
end = timer()
print(end - start) # time in seconds
```
For each variation:
```python
6.0040998505428433e-05 # original
0.00016220801626332104 # this commit
0.00014566699974238873 # leaving the function in _record_agents
4.683298175223172e-05 # leaving the function in _record_agents and removing the model argument in _record_agents that is no longer used
```

So actually I was mistaken but we can get it to be pretty fast by leaving the func inside and removing the unused arg. The functions in question for the last timed test are:
```python
def _record_agents(self, schedule):
    """Record agents data in a mapping of functions and agents."""
    rep_funcs = self.agent_reporters.values()

    def get_reports(agent):
        _prefix = (schedule.steps, agent.unique_id)
        reports = tuple(rep(agent) for rep in rep_funcs)
        return _prefix + reports

    agent_records = map(get_reports, schedule.agents)
    return agent_records

def collect(self, model):
    """Collect all the data for the given model object."""
    if self.schedule is None:
        schedule = model.schedule
    else:
        schedule = self.schedule

    if self.model_reporters:

        for var, reporter in self.model_reporters.items():
            # Check if Lambda operator
            if isinstance(reporter, types.LambdaType):
                self.model_vars[var].append(reporter(model))
            # Check if model attribute
            elif isinstance(reporter, partial):
                self.model_vars[var].append(reporter(model))
            # Check if function with arguments
            elif isinstance(reporter, list):
                self.model_vars[var].append(reporter[0](*reporter[1]))
            else:
                self.model_vars[var].append(self._reporter_decorator(reporter))

    if self.agent_reporters:
        agent_records = self._record_agents(schedule)
        self._agent_records[schedule.steps] = list(agent_records)
```

Thoughts?

Reply:
I think having the function defined inside `_record_agents` is clearer. This way, you can have the name remain short, `get_reports`.